rbi/annotations/activejob.rbi

```diff
@@ -0,0 +1,90 @@
+# typed: true
+
+class ActiveJob::Base
+  sig { params(blk: T.proc.bind(T.attached_class).params(job: T.attached_class).void).void }
+  def self.after_discard(&blk); end
+
+  sig do
+    params(filters: T.untyped, blk: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class).void)).void
+  end
+  def self.after_enqueue(*filters, &blk); end
+
+  sig do
+    params(filters: T.untyped, blk: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class).void)).void
+  end
+  def self.after_perform(*filters, &blk); end
+
+  sig do
+    params(
+      filters: T.untyped,
+      blk: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class, block: T.untyped).void)
+    ).void
+  end
+  def self.around_enqueue(*filters, &blk); end
+
+  sig do
+    params(
+      filters: T.untyped,
+      blk: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class, block: T.untyped).void)
+    ).void
+  end
+  def self.around_perform(*filters, &blk); end
+
+  sig do
+    params(filters: T.untyped, blk: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class).void)).void
+  end
+  def self.before_enqueue(*filters, &blk); end
+
+  sig do
+    params(filters: T.untyped, blk: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class).void)).void
+  end
+  def self.before_perform(*filters, &blk); end
+
+  sig do
+    type_parameters(:ExceptionType)
+      .params(
+        exceptions: T::Class[T.type_parameter(:ExceptionType)],
+        block: T.nilable(
+          T.proc.bind(T.attached_class).params(
+            job: T.attached_class,
+            error: T.type_parameter(:ExceptionType)
+          ).void
+        )
+      ).void
+  end
+  sig do
+    params(
+      exceptions: String,
+      block: T.nilable(T.proc.bind(T.attached_class).params(job: T.attached_class, error: T.untyped).void),
+    ).void
+  end
```

Comment:
👋 Hey Bart, it's been a while. Hope you're well! Jumping in here as it seems a bit off that `discard_on` is RBI typed differently than `rescue_on` and `retry_on` considering according to the Rails docs they're all the same.

- https://api.rubyonrails.org/classes/ActiveJob/Exceptions/ClassMethods.html#method-i-discard_on
- https://api.rubyonrails.org/classes/ActiveJob/Exceptions/ClassMethods.html#method-i-retry_on
- https://api.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html#method-i-rescue_from

Thoughts on bringing `discard_on` in-line with the others?
- https://github.com/Shopify/rbi-central/pull/263

Bonus, would allow `discard_on` usage to avoid needing `T.unsafe` with the latest Sorbet provided the array of strings is frozen via `freeze` re: the following change.
- https://github.com/sorbet/sorbet/pull/7993

Reply:
Hey @larouxn 😄 hope you're doing well too!

You're correct they're all the same but the `discard_on` sigs are more accurate when the first sig with generics applies. `rescue_on` and `retry_on` should ideally follow this example, except for the fact that we currently can't use overloads because of keyword arguments: https://github.com/sorbet/sorbet/issues/37

Reply:
Understood. Assuming Sorbet keyword argument support is fixed, we should update all to use the more accurate overload style approach. Assuming support is _not_ fixed, current state, does it make sense to bring `discard_on` in line with the other two? In the end, I'm thinking consistency either way would be best, leaning towards non-overload given we don't know when if ever kwarg support will be fixed. 😅

Reply:
In my experience working with latest Sorbet, as of https://github.com/sorbet/sorbet/pull/7993, splatting a frozen array of strings with `rescue_on` and `retry_on` is fine, no errors but doing the same with `discard_on` will lead to an error due to the stricter expectation that all string/non-string classes passed to it are of `ExceptionType`.

```ruby
Expected T::Class[T.type_parameter(:ExceptionType)] but found T.class_of(Permanent) for argument exceptions https://srb.help/7002
```

Thus, it's not a bug, it's expected behaviour. But at the same time, I think it would be clearer if all three ActiveJob methods had the same typing/expectations.

Reply:
Update, seemingly good news, appears we can't leave `ExceptionType` as is and update the signature below from `exceptions` → `String` to `T.any(Module, String)`, like the other two methods. https://github.com/Shopify/rbi-central/pull/263#issuecomment-2249140232 🙌